src/main/ruby/truffleruby/core/regexp.rb

```diff
@@ -55,22 +55,23 @@ def self.try_convert(obj)
     Truffle::Type.try_convert obj, Regexp, :to_regexp
   end
 
-  def self.convert(pattern)
-    return pattern if Primitive.is_a?(pattern, Regexp)
-    if Primitive.is_a?(pattern, Array)
-      union(*pattern)
-    else
-      Regexp.quote(pattern.to_s)
-    end
-  end
+  def self.negotiate_union_encoding(*patterns)
+    res = nil
+
+    patterns.each do |pattern|
+      converted = Primitive.is_a?(pattern, Regexp) ? pattern : Regexp.quote(pattern)
 
-  def self.compatible?(*patterns)
-    encodings = patterns.map { |r| convert(r).encoding }
-    last_enc = encodings.pop
-    encodings.each do |encoding|
-      raise ArgumentError, "incompatible encodings: #{encoding} and #{last_enc}" unless Primitive.encoding_compatible?(last_enc, encoding)
-      last_enc = encoding
+      enc = converted.encoding
+
+      if Primitive.nil?(res)
+        res = enc
+      else
+        res = Primitive.encoding_compatible?(enc, res)
+        raise ArgumentError, "incompatible encodings: #{enc} and #{res}" unless res
```

Comment:
`res` in the exception message will always be `nil` here, we need to save the value in a local variable. I'll do it.

Reply:
Good catch. Thanks. I had originally written it using `reduce` and missed this when I restructured the code.